Gutenboarding: v2 improvements for header and domain search

razvanpapadopol opened the pull request:
### Changes proposed in this Pull Request

**Domain nudge and header updates**
* Show correct states of Domain Button with correct data fixing issues mentioned in https://github.com/Automattic/wp-calypso/pull/44655#issuecomment-671375706
* Refactor header and extract `DomainPickerButton` to load domain suggestion only after mounting the component and only if there is no selected domain. This means no more domain requests on Intent Capture step and also removing `useDebounce` dependency.

**Animation**
* Re-add loading status of domain button. 
* Re-add slide-in animation to highlight domain button until Domain Picker is used.
Demo: https://cloudup.com/c7ceOH98tvx

**Other**
* Use vertical as fallback for DomainPicker initial input value.
* Hide Plans button on Plans step and PlansModal (is causing a navigation bug).

### Testing instructions
* Go to /new
* Test all the possible cases described above
  * with a value for site title / vertical
  * with no value on intent capture step
  * with a changed value
  * before / after interacting with the domain picker (custom query or selecting a domain)

Fixes part of #44655

simison commented on client/landing/gutenboarding/components/domain-picker-button/index.tsx:
```diff
@@ -3,41 +3,83 @@
+	const domainSuggestion = useSelect(
+		( select ) => {
+			// Get suggestion only if the query is valid and if there isn't a selected domain
+			if ( ! suggestionQuery || suggestionQuery.length < 2 || domain ) {
```
Testing that `suggestionQuery` still has something if everything else but alpha-numerics are removed would be nice addition, but nothing urgent. You can try emojis or some Japanese for example. :-)

razvanpapadopol replied:
You mean [this issue](https://github.com/Automattic/wp-calypso/issues/40748)? 😄  
I'd be happy to take it in the next sprint as a follow up of current update.

simison commented on client/landing/gutenboarding/components/domain-picker-button/style.scss:
```diff
@@ -1,24 +1,44 @@
+
+	&--has-placeholder {
+		@include onboarding-placeholder;
+		@include reduce-motion( 'animation' );
```
Shouldn't the `onboarding-placeholder` just directly include `reduce-motion`? Not a biggie for this PR.

simison commented:
> Re-add loading status of domain button.

Did we have this at some point? 🤔 

![Screen Recording 2020-08-12 at 19 52 47](https://user-images.githubusercontent.com/87168/90044154-e963d480-dcd5-11ea-9113-69a7b8aecd89.gif)


I don't think we should have it pulsating and then appear animated again, in my memory previously there wasn't first anything and then it appeared a bit delayed sometime around design picker. cc @olaolusoga @ollierozdarz for comment.

That's how it's also in v2 recording video (find it from pbAok1-16b-p2)

simison commented:
I get the animation appear again and again between steps in mobile flow. On desktop no such issue.

ollierozdarz commented:
> > Re-add loading status of domain button.
> 
> Did we have this at some point? 🤔 In this PR:
> 
> ![Screen Recording 2020-08-12 at 19 52 47](https://user-images.githubusercontent.com/87168/90044154-e963d480-dcd5-11ea-9113-69a7b8aecd89.gif)
> 
> I don't think we should have it pulsating and then appear animated again, in my memory previously there wasn't first anything and then it appeared a bit delayed sometime around design picker. cc @olaolusoga @ollierozdarz for comment.
> 
> This is from v2 recording video (find it via pbAok1-16b-p2):
> 
> ![Screen Recording 2020-08-12 at 19 57 35](https://user-images.githubusercontent.com/87168/90044414-46f82100-dcd6-11ea-8df1-f5f06517a336.gif)

I believe @dubielzyk created this original animation and could probably answer best the original intention of implementation. Looking with fresh eyes, I don't think it should appear pulse in and continue to animate again, it detracts from the smooth easing of it (@simison's thoughts also I think)

dubielzyk commented:
The original idea was that the animation is pretty much as the v2 recording video, but it should've appeared as you land on the design step (after intent and before fonts). I think the way it's implemented is that it's just on a timer, but I'm unsure. It definitely doesn't seem to be consistent though. Happy to do what @ollierozdarz believes is best :)

razvanpapadopol commented:
Removed the pulse animation and adjusted the slide-in to happen only if use didn't selected domain, design or entered a domain search query. So basically you'll get the slide-in only if you navigate just between intent capture step and design selection step. After that, when loading new suggestion it will have the standard subtle opacity transition we're using (200ms).

Demo: https://cloudup.com/cMpRgKrSOLc

yansern commented on client/landing/gutenboarding/components/header/index.tsx:
```diff
@@ -28,61 +25,21 @@ const Header: React.FunctionComponent = () => {
-				</span>
-		  );
+	// steps (including modals) where we show Domains button
+	const showDomainsButton = [ 'DesignSelection', 'Style', 'Features', 'Plans' ].includes(
```
Would be nice if these were constants instead of string. Looks like we don't have constants for step keys (or `StepNameType`) in `path.ts`. Nice-to-have for future PR.

simison replied:
Could be good simple dev-onboarding task :-)

yansern commented on client/landing/gutenboarding/components/header/index.tsx:
```diff
@@ -28,61 +25,21 @@ const Header: React.FunctionComponent = () => {
-		'CreateSite',
-	].includes( currentStep );
+	// steps (including modals) where we hide Plans button
+	const showPlansButton = [ 'DesignSelection', 'Style', 'Features' ].includes( currentStep );
```
Same with this one. https://github.com/Automattic/wp-calypso/pull/44864/files#r469918625

yansern commented:
Not a big issue. Clicking on a domain item and then clicking "Go back" would still use that domain as the selected domain. So there's actually no functional difference with clicking "Continue". I'd expect "Go back" to function more like cancel, but perhaps that could be argued. Fine for PR to proceed.
